File: distinct_subsequences.py
class Solution:
    def numDistinct(self,S,T):
        m=len(S)
        n=len(T)
        dp=[[0 for j in range(m+1)] for i in range(n+1)]
        for j in range(m+1):
            dp[0][j]=1 # T is null
        for i in range(1,n+1):
            for j in range(1,m+1):
                dp[i][j]=dp[i][j-1]
                if S[j-1]==T[i-1]:
                    dp[i][j]=dp[i-1][j-1]+dp[i][j-1]
                   
        return dp[n][m]
    def numDistinct1(self,S,T):
        m=len(S)
        n=len(T)
        dp=[[0 for j in range(n+1)] for i in range(m+1)]
        for i in range(m+1):
            dp[i][0] = 1 # T is null
        for i in range(1,m+1):
            for j in range(1,n+1):
                dp[i][j]=dp[i-1][j]
                if S[i-1]==T[j-1]:
                    dp[i][j]=dp[i-1][j-1]+dp[i-1][j]
                   
        return dp[m][n]        

if __name__=='__main__':
    sol=Solution()
    S='rabbbit'
    T='rabbit'
    print sol.numDistinct(S,T)
    print sol.numDistinct1(S,T)

File: Algorithm/CPP/distinct_subsequences.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
using namespace std;

class Solution {
public:
	int numDistinct1(string s, string t) {
		int m = t.size();
		int n = s.size();
		if (m > n) return 0;
		vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));
		for (int i = 0; i <=n; i++)
			dp[0][i] = 1;
		for (int j = 0; j < n; j++)
			for (int i = 0; i < m; i++)
				dp[i + 1][j + 1] =  dp[i + 1][j] + (s[j] == t[i] ? dp[i][j] : 0);
		return dp[m][n];
	}
	int numDistinct(string s, string t) {
		int m = t.size();
		int n = s.size();
		if (m > n) return 0;
		vector<int> dp(m+1,0);
		dp[0] = 1;
		for (int j = 0; j < n; j++)
		{
			for (int i = min(j, m - 1); i >= 0; i--)
				dp[i + 1] = dp[i + 1] + (s[j] == t[i] ? dp[i] : 0);
		}
		return dp[m];
	}
		int numDistinct3(string s, string t) {
		int m = t.size();
		int n = s.size();
		if (m > n) return 0;
		vector<int> dp(m + 1, 0);
		dp[0] = 1;
		for (int j = 0; j < n; j++)
		{
			for (int i = min(j, m - 1); i >= 0; i--)
				dp[i + 1] = dp[i + 1] + (s[j] == t[i] ? dp[i] : 0);
		}
		return dp[m];
	}
};


int main()
{
	Solution sol;
	vector<int> nums = {1,2,3,1};
	string s = "ccc";
	string t = "cc";
	cout << sol.numDistinct(s,t);

	cout << endl;
	//for (auto k :nums )cout << k << " ";
	system("pause");
	return 0;
}
